neqo-crypto/src/aead.rs

```diff
@@ -157,10 +194,33 @@ impl Aead {
         }?;
         Ok(&output[0..(l.try_into()?)])
     }
+
+    #[allow(clippy::unused_self)]
+    fn fuzzing_mode(&self) -> bool {
+        false
+    }
 }
 
-impl fmt::Debug for Aead {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "[AEAD Context]")
+/// Create a new AEAD based on the fuzzing mode, indicated TLS version and
+/// cipher suite. When fuzzing is enabled, return `AeadFuzzing`.
+///
+/// # Errors
+/// Returns `Error` when the supporting NSS functions fail.
+pub fn create_aead(
+    #[cfg(feature = "fuzzing")] fuzzing: bool,
+    #[cfg(not(feature = "fuzzing"))] _fuzzing: bool,
+    version: Version,
+    cipher: Cipher,
+    secret: &SymKey,
+    prefix: &str,
+) -> Res<Box<dyn Aead>> {
+    #[cfg(feature = "fuzzing")]
+    if fuzzing {
+        return Ok(Box::new(
+            AeadFuzzing::new(version, cipher, secret, prefix).unwrap(),
+        ));
     }
+
+    let aead = AeadImpl::new(version, cipher, secret, prefix)?;
+    Ok(Box::new(aead))
```

Comment:
This is why I think that we should avoid the trait.  This isn't a huge deal, but the box isn't needed.

Let's say that you have fuzz::Aead with the defined interface (not a trait, just the necessary functions).  This is instantiated with `Option<real::Aead>`, such that if that is present, the various functions call back to the real function.

Then this module looks like this:

```rust
#[cfg(feature="fuzzing")]
pub use ...::fuzzing_aead::FuzzingAead;

#[cfg(not(feature="fuzzing"))]
pub use RealAead as Aead;

// ... 

impl RealAead {
pub fn new(...)
  // ...
}
}
```

The fuzzing AEAD can look like this:

```rust
pub struct FuzzingAead {
  real: Option<RealAead>;
}

pub fn new(...) {
    let real = if fuzzing {
      None
    } else {
       Some(RealAead::new(version, cipher, secret, prefix)?)
    };
    Self::new(real, version, cipher, secret, prefix)?
}
```

The advantage with that arrangement is that it ensures that you don't have a dynamic dispatch every time the AEAD is invoked (which is a fair bit).  The fuzzing feature will change what the rest of the code thinks is called "Aead", so that it ends up having different code when the feature is enabled.